ddcloud/resource_server.go

```diff
@@ -488,6 +497,28 @@ func resourceServerUpdate(data *schema.ResourceData, provider interface{}) error
 		}
 	}
 
+	if data.HasChange(resourceKeyServerPowerState) {
+		log.Printf("Server power state change has been detected.")
+		powerState := propertyHelper.GetOptionalString(resourceKeyServerPowerState, false)
+
+		switch strings.ToLower(*powerState) {
+		case "start":
+			err = serverStart(providerState, serverID)
+		case "shutdown":
+			err = serverShutdown(providerState, serverID)
+		case "off":
+			err = serverPowerOff(providerState, serverID)
+		default:
+			err = fmt.Errorf("Invalid power State (%s); Valid Power states are start, shutdown, off", *powerState)
```

Comment:
This will panic if `powerState` is `nil`, consider ignoring power state if it's `nil` since that means the user hasn't specified it (maybe see how disks are handled for an example of this behaviour?).

Reply:
More broadly, we should probably discuss how this affects existing users; perhaps provide additional values so users have to opt-in to this behaviour? For example, if I don't specify `power_state` or set it to `autostart` then it only affects newly-deployed servers (i.e. ignored during subsequent run of `terraform apply`).

Reply:
Correct me if i'm wrong.
If power state hasn't been specified. It'll use the default out of the resourceKeyServerPowerState schema and never be nil?

However in saying that I would prefer not to have any code that could panic and will put in a 
if powerState != nil {
}
catch

Reply:
If it's not optional (i.e. can never be `nil`) then you can just use `GetString` instead of `GetOptionalString`.

Reply:
(yes, looking more closely at the code you're correct and it can't be `nil`)

Reply:
I added a nil check to the pointer, in case of the event someone somehow managed to do something out of the norm.